source/vstd/std_specs/hash.rs

```diff
@@ -0,0 +1,383 @@
+// model that type here. In particular, we have an axiom that
+// `RandomState` conforms to our model of how `BuildHasher` behaves.
+#[verifier::external_type_specification]
+#[verifier::external_body]
+pub struct ExRandomState(RandomState);
+
+pub broadcast proof fn axiom_random_state_conforms_to_build_hasher_model()
+    ensures
+        does_type_conform_to_build_hasher_model::<RandomState>(),
+{
+    admit();
+}
+
+// We now specify the behavior of `HashMap`.
+#[verifier::external_type_specification]
+#[verifier::external_body]
+#[verifier::reject_recursive_types(Key)]
+#[verifier::reject_recursive_types(Value)]
+#[verifier::reject_recursive_types(S)]
+pub struct ExHashMap<Key, Value, S>(HashMap<Key, Value, S>);
+
+pub trait HashMapAdditionalSpecFns<Key, Value>: View<V = Map<Key, Value>> {
+    spec fn spec_index(&self, k: Key) -> Value
+        recommends
+            self@.contains_key(k),
+    ;
+}
+
+impl<Key, Value, S> HashMapAdditionalSpecFns<Key, Value> for HashMap<Key, Value, S> {
+    #[verifier::inline]
+    open spec fn spec_index(&self, k: Key) -> Value {
+        self@.index(k)
+    }
+}
+
+impl<Key, Value, S> View for HashMap<Key, Value, S> {
+    type V = Map<Key, Value>;
+
+    #[verifier::external_body]
+    closed spec fn view(&self) -> Map<Key, Value>;
+}
+
+pub open spec fn spec_hash_map_len<Key, Value, S>(m: &HashMap<Key, Value, S>) -> usize;
+
+pub broadcast proof fn axiom_spec_hash_map_len<Key, Value, S>(m: &HashMap<Key, Value, S>)
+    ensures
+        does_type_have_deterministic_hash::<Key>() && does_type_conform_to_build_hasher_model::<S>()
+            ==> #[trigger] spec_hash_map_len(m) == m@.len(),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+#[verifier::when_used_as_spec(spec_hash_map_len)]
+pub fn ex_hash_map_len<Key, Value, S>(m: &HashMap<Key, Value, S>) -> (len: usize)
+    ensures
+        len == spec_hash_map_len(m),
+{
+    m.len()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_map_new<Key, Value>() -> (m: HashMap<Key, Value, RandomState>)
+    ensures
+        m@ == Map::<Key, Value>::empty(),
+{
+    HashMap::<Key, Value>::new()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_map_with_capacity<Key, Value>(capacity: usize) -> (m: HashMap<
+    Key,
+    Value,
+    RandomState,
+>)
+    ensures
+        m@ == Map::<Key, Value>::empty(),
+{
+    HashMap::<Key, Value>::with_capacity(capacity)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_map_reserve<Key, Value, S>(m: &mut HashMap<Key, Value, S>, additional: usize) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        m@ == old(m)@,
+{
+    m.reserve(additional)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_map_insert<Key, Value, S>(
+    m: &mut HashMap<Key, Value, S>,
+    k: Key,
+    v: Value,
+) -> (result: Option<Value>) where Key: Eq + Hash, S: BuildHasher
+    ensures
+        m@ == old(m)@.insert(k, v),
```

Comment:
Should this equality also be gated on `does_type_have_deterministic_hash::<Key>() && does_type_conform_to_build_hasher_model::<S>()`?

Reply:
Also, does it account for this bit of the [documentation](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert:~:text=The%20key%20is%20not%20updated%2C%20though%3B%20this%20matters%20for%20types%20that%20can%20be%20%3D%3D%20without%20being%20identical)?
> The key is not updated, though; this matters for types that can be == without being identical

Reply:
I figured the inequality didn't need to be gated on those, because it only affects the abstract view of the map, which we can consider to have changed even if there's no way to get concrete access to that abstract view. I guess this is a pretty subtle point, though, and I might as well gate it.

As for that bit of documentation, no, I hadn't taken it into account! I'll have to change `does_type_have_deterministic_hash` to something stronger implying that it not only has a deterministic hash but is suitable for this model of HashMap. Specifically, it can't do those shenanigans with two non-identical keys satisfying `==`.

Eventually, I also plan to write a wrapper named `HashMapWithView` or something that represents a `HashMap` whose `Key` satisfies `View<KeyView>`. Then, it'll be abstracted as a `Map<KeyView, Value>`. But I think `HashMap` should stick with only modeling "normal" keys.

Reply:
I've made the changes.

Reply:
Thanks very much!